src/main/java/uk/gov/companieshouse/overseasentitiesapi/interceptor/UserAuthenticationInterceptor.java

```diff
@@ -74,4 +103,82 @@ public boolean preHandle(@NonNull HttpServletRequest request, @NonNull HttpServl
+                    authInfoMap);
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return false;
+        }
+    }
+
+    private boolean isRequestRoeUpdate(String companyNumberInScope) {
+        // Further check for OE number is done in doRoeUpdateJourneyChecks method
+        return StringUtils.isNotEmpty(companyNumberInScope);
+    }
+
+    private String getCompanyNumberInScope (HttpServletRequest request) {
+        final Map<String, List<String>> privileges = getERICTokenPermissions(request);
+        if (privileges.containsKey(COMPANY_NUMBER_KEY)) {
+            return privileges.get(COMPANY_NUMBER_KEY).get(0);
+        }
+        return null;
+    }
+
+    private boolean doCompanyNumbersMatch(HttpServletRequest request, HttpServletResponse response, String companyNumberInScope, String transactionId) {
+        String companyNumberInTransaction = getCompanyNumberInTransaction(request, response, transactionId);
+        if (TRANSACTION_SERVER_ERROR.equals(companyNumberInTransaction)) {
+            return false;
+        }
+        if (StringUtils.isNotBlank(companyNumberInTransaction) && companyNumberInTransaction.equalsIgnoreCase(companyNumberInScope)) {
+            return true;
+        }
+        response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+        return false;
+    }
+
+    private Map<String, List<String>> getERICTokenPermissions(HttpServletRequest request) {
+        String tokenPermissionsHeader = request.getHeader(ERIC_AUTHORISED_TOKEN_PERMISSIONS);
+        Map<String, List<String>> permissions = new HashMap<>();
+        if (tokenPermissionsHeader != null) {
+            for (String pair : tokenPermissionsHeader.split(" ")) {
+                String[] parts = pair.split("=");
+                permissions.put(parts[0], Arrays.asList(parts[1].split(",")));
+            }
+        }
+        return permissions;
+    }
+
+    private String getCompanyNumberInTransaction(HttpServletRequest request, HttpServletResponse response, String transactionId) {
+        String passthroughHeader = request.getHeader(ApiSdkManager.getEricPassthroughTokenHeader());
+        String reqId = request.getHeader(ERIC_REQUEST_ID_KEY);
+        var logMap = new HashMap<String, Object>();
+        logMap.put(TRANSACTION_ID_KEY, transactionId);
+        try {
+            final var transaction = transactionService.getTransaction(transactionId, passthroughHeader, reqId);
+            logMap.put(COMPANY_NUMBER_KEY, transaction.getCompanyNumber());
+            ApiLogger.debugContext(reqId, "Transaction successfully retrieved " + transactionId, logMap);
+            return transaction.getCompanyNumber();
+        } catch (Exception e) {
+            ApiLogger.errorContext(reqId, "Error retrieving transaction " + transactionId, e, logMap);
+            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+            return TRANSACTION_SERVER_ERROR;
```

Comment:
Instead of returning a special value consider returning Optional<String>. The caller can check for Optional.isEmpty() instead of the special value.

Reply:
Yes, I agree, using the string return value to influence behaviour is not great - either it contains a company number or it doesn't, so yes, making use of Optional here is a good option I think.

Reply:
would need Optional<Optiona<String>>